services/blockstorage/adapter/filesystem/index.go

```diff
@@ -64,54 +71,73 @@ func (i *blockHeightIndex) fetchBlockOffset(height primitives.BlockHeight) (offs
 	return
 }
 
+// assumes timeStamp(inOrderBlock) <= timeStamp(topBlock)
 func (i *blockHeightIndex) getEarliestTxBlockInBucketForTsRange(rangeStart primitives.TimestampNano, rangeEnd primitives.TimestampNano) (primitives.BlockHeight, bool) {
 	i.RLock()
 	defer i.RUnlock()
 
 	fromBucket := blockTsBucketKey(rangeStart)
 	toBucket := blockTsBucketKey(rangeEnd)
+	inOrderHeight := getBlockHeight(i.inOrderBlock)
 	for b := fromBucket; b <= toBucket; b++ {
-		result, exists := i.firstBlockInTsBucket[b]
-		if exists {
-			return result, true
+		blockHeight, exists := i.firstBlockInTsBucket[b]
+		if blockHeight > inOrderHeight {
+			return 0, false
+		} else if exists {
+			return blockHeight, true
 		}
 	}
 	return 0, false
 
 }
 
+func (i *blockHeightIndex) validateCandidateBlockHeight(candidateBlockHeight primitives.BlockHeight) error {
+	i.RLock()
+	defer i.RUnlock()
+
+	topHeight := getBlockHeight(i.topBlock)
+	inOrderHeight := getBlockHeight(i.inOrderBlock)
+
+	if (i.lastSyncedHeight > inOrderHeight && candidateBlockHeight != i.lastSyncedHeight-1) ||
```

Comment:
this looks correct but its a little hard to read. consider breaking down to two separate conditions like:

```
var result error
if (i.lastSyncedHeight > inOrderHeight && candidateBlockHeight != i.lastSyncedHeight-1)) {

  error = fmt.Error("sync session in progress, expected block height %d", i.lastSyncedHeight-1)

} else if (inOrderHeight == topHeight && candidateBlockHeight <= inOrderHeight) {

  error = fmt.Error("expected block height higher than current top %d",  inOrderHeight)

}
return error
```